Minikore outer

lucaspena opened the pull request:
@daejunpark @msaxena2 @cos
Please review, don't merge.

First rough cut for interface for outer kore.
Made outer.scala files at implementation and interfaces level.
Put basic derived operations at trait level for Definition, Module, Sentence.
Can't build Definition, Module, Sentence abstractly, currently need to rely on implementation. Perhaps build methods can be added to the traits to fix this.

daejunpark commented on src/main/scala/org/kframework/minikore/implementation/MiniKore.scala:
```diff
@@ -1,122 +1,163 @@
 
-  type Attributes = Seq[i.Pattern]
+  // Pattern Implementations
+  // =======================
```
This diff is not readable. Could you make it to have a more meaningful diff? Otherwise, please briefly explain what was changed.

daejunpark commented on src/main/scala/org/kframework/minikore/interfaces/outer.scala:
```diff
@@ -0,0 +1,67 @@
+
+  import org.kframework.minikore.interfaces.{pattern => p}
+
+  type Attributes = Seq[p.Pattern]
```
The same type declaration in both here and the pattern interface?

ehildenb replied:
I don't think it's also in the Pattern interface, is it?

daejunpark commented on src/main/scala/org/kframework/minikore/interfaces/outer.scala:
```diff
@@ -0,0 +1,67 @@
+
+  trait Sentence extends HasAttributes {
+    val sorts: Set[p.Sort] = Set.empty
+    val symbols: Set[p.Symbol] = Set.empty
```
What's the use of `sorts` and `symbols` for the sentences?

daejunpark replied:
I know why you put the empty set as the default value. But it was confusing when I read the interface first. I think it makes more sense to have these not implemented here.

msaxena2 replied:
Should we try to a figure out a way to not have sort and symbols in the Sentence trait, given that not all Sentences have Sorts/Symbols?

daejunpark commented on src/main/scala/org/kframework/minikore/interfaces/outer.scala:
```diff
@@ -0,0 +1,67 @@
+    val sorts: Set[p.Sort] = Set.empty
+    val symbols: Set[p.Symbol] = Set.empty
+
+    override def onAttributes(f: p.Pattern => p.Pattern): Sentence
```
By having the build method, we can implement this method here.

daejunpark commented on src/main/scala/org/kframework/minikore/interfaces/outer.scala:
```diff
@@ -0,0 +1,67 @@
+
+    override def onAttributes(f: p.Pattern => p.Pattern): Sentence
+  }
+
```
I think  we should have the interfaces for each sentence types, `Import`, `Rule`, `Axiom`, etc.

msaxena2 replied:
I agree with @daejunpark on this.

daejunpark commented:
Overall, it makes sense. It seems reasonable to add the `sorts` and `symbols` members.

Suggestion:
 * Add interface for each sentence types, and implement `onAttributes` method there.
 * Put back outer.scala in MiniKore.scala.

msaxena2 commented on src/main/scala/org/kframework/minikore/implementation/MiniKore.scala:
```diff
@@ -1,122 +1,163 @@
+}
+
+
+/** Helpers for building MiniKore definitions **/
```
This should be moved to a separate file, in my opinion. This object should be paramteric over a builder interface, that builds outer structures.

msaxena2 commented on src/main/scala/org/kframework/minikore/implementation/outer.scala:
```diff
@@ -0,0 +1,45 @@
+/**
+  * Created by lpena on 3/23/17.
+  */
+object outer {
```
outer should extend something an OuterBuilders trait, like DefaultBuilders extends the Builders trait for pattern

msaxena2 commented on src/main/scala/org/kframework/minikore/interfaces/outer.scala:
```diff
@@ -0,0 +1,67 @@
+
+  type Attributes = Seq[p.Pattern]
+
+  trait HasAttributes {
```
I think we can just call this the Attributes trait, instead of calling it HasAttributes. We'd like this functionality over attributes in general right?

ehildenb replied:
I'm worried about the name clash between `trait Attributes` and `type Attributes = Seq[p.Pattern]` being confusing. Thoughts?

msaxena2 commented on src/main/scala/org/kframework/minikore/interfaces/outer.scala:
```diff
@@ -0,0 +1,67 @@
+    val modules: Seq[Module]
+
+    // Derived operations
+    val sorts: Set[p.Sort] = modules flatMap (_.sorts) toSet
```
We took a decision to avoid the dotless notation. I think we should keep it that way for sake of consistency.

msaxena2 commented on src/main/scala/org/kframework/minikore/interfaces/outer.scala:
```diff
@@ -0,0 +1,67 @@
+    // Derived operations
+    val sorts: Set[p.Sort] = modules flatMap (_.sorts) toSet
+    val symbols: Set[p.Symbol] = modules flatMap (_.symbols) toSet
+    val sentences: Seq[Sentence] = modules flatMap (_.sentences)
```
These should be lazy vals, right?

daejunpark commented:
Please let me know (by leaving a message here with mentioning my name) once it is ready to review, so that I can quickly review and we can merge.

grosu commented:
Same here, let me know when I can review it.  We should merge this asap, especially because it is not difficult.  The difficult part is probably to ensure test coverage.  Let us please give this some priority.

lucaspena commented:
Please take a look at the changes we made. We addressed your comments from
before, and did some refactoring as well. We still have to add
comments/tests, and make K work with the refactoring.

On Mon, Apr 3, 2017 at 8:02 PM, Daejun Park <notifications@github.com>
wrote:

> Please let me know (by leaving a message here with mentioning my name)
> once it is ready to review, so that I can quickly review and we can merge.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kframework/kore/pull/15#issuecomment-291355354>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AD1ZLk3-BvjftoAmGRPx_57BZm9R2BDHks5rsZahgaJpZM4Mncyb>
> .
>



-- 
Lucas Peña
University of Illinois at Urbana-Champaign
lucaspena13@gmail.com
(954) 882-7070

daejunpark commented on src/main/scala/org/kframework/kore/KoreDSL.scala:
```diff
@@ -0,0 +1,58 @@
+  }
+  implicit def asImport(i: imports): o.Import = i.att()
+
+  case class sort(sort: p.Sort) {
```
sortDecl or declSort would be better. I was confused with the one corresponding to Sort.

daejunpark commented on src/main/scala/org/kframework/kore/KoreDSL.scala:
```diff
@@ -0,0 +1,58 @@
+  }
+  implicit def asSortDeclaration(s: sort): o.SortDeclaration = s.att()
+
+  case class symbol(sort: p.Sort, symbol: p.Symbol, args: p.Sort*) {
```
same here. declSymbol or symbolDecl would be better.

daejunpark commented on src/main/scala/org/kframework/kore/KoreDSL.scala:
```diff
@@ -0,0 +1,58 @@
+
+  // Why have both Rule and Axiom?
+  case class rule(l: p.Pattern, r: p.Pattern) {
+    def att(atts: p.Pattern*): o.Axiom = ob.Axiom(pb.Rewrite(l, r), atts)
```
Should be Rule instead of Axiom.

daejunpark commented on src/main/scala/org/kframework/kore/KoreDSL.scala:
```diff
@@ -0,0 +1,58 @@
+  }
+  implicit def asAxiom(a: axiom): o.Axiom = a.att()
+
+  // Why have both Rule and Axiom?
```
I explained it last time. The transition system induced by the definition is:
``` 
\big_or( rules )  \and  \big_and( axioms )
```
Maybe it is a good idea to put it as a comment (in some proper place) once you understand it.

daejunpark commented on src/main/scala/org/kframework/kore/KoreDSL.scala:
```diff
@@ -0,0 +1,58 @@
+  case class rule(l: p.Pattern, r: p.Pattern) {
+    def att(atts: p.Pattern*): o.Axiom = ob.Axiom(pb.Rewrite(l, r), atts)
+  }
+  implicit def asAxiom(r: rule): o.Axiom = r.att()
```
Should be asRule.

daejunpark commented on src/main/scala/org/kframework/kore/KoreDSL.scala:
```diff
@@ -0,0 +1,58 @@
+package org.kframework.kore
```
How about creating a new directory (say `util`) and put this file there? I guess we will add more like this.

daejunpark commented on src/main/scala/org/kframework/kore/KoreDSL.scala:
```diff
@@ -0,0 +1,58 @@
+  //implicit def asValue(s: String): p.Value = pb.Value(s)
+
+  case class definition(modules: o.Module*) {
+    def att(atts: p.Pattern*): o.Definition = ob.Definition(modules, atts)
```
Is not it necessary to put *.toSeq for modules and atts, like: ```ob.Definition(modules.toSeq, arts.toSeq)```?

daejunpark commented on src/main/scala/org/kframework/kore/interfaces/Interfaces.scala:
```diff
@@ -481,61 +481,183 @@ object pattern {
+    def Rewrite(_1: Pattern, _2: Pattern): Rewrite
+    def Application(_1: Symbol, args: Seq[Pattern]): Application
+  }
+
```
Can we flatten OuterBuilders and PatternBuilders and just have builders?

lucaspena replied:
There's a separation between outer and pattern objects, so I think it makes sense to have a separation between the outerbuilders and patternbuilders objects. Is there a reason you think they should be flattened?

cos replied:
@daejunpark, if we merge them, we shoot ourselves in the foot regarding being able to create a definition using default pattern builders, then be able to create specialized pattern builders from it, something I know you agree of.

grosu replied:
@cos Why?  We need anyway some utility function that takes a KORE definition (using any builder of patterns as well as of outer, default or not) and yields a definition using any other given builder, don't we?  Basically it would just visit it using the KORE extractors and build the new one at the same time using the second builder.  Non-uniformity is not good in general.  Splitting the KORE interface in two parts and handling them differently feels wrong to me.  There is no separation in the KORE language, so there should be no separation in the interface either.

cos replied:
This is related to what Daejun was arguing about in a recent K meeting, and I said we can make it work. Yes, we can make it work, but this separation still exists, and I think we should leave in the interfaces.

Some backends (e.g., both the Java one and Kale) can only construct correct pattern objects once they have seen the entire definition -- this is natural because some symbols only make sense once several sentences have been seen, e.g., you can only construct an assoc once you can see the symbol for its identity element. 

E.g.,
```
val kale = new KaleBuilders()
val setPatternDeclaredVeryEarly = kale.Application(Symbol("_ _"), Seq(1, 2)) // 1 and 2 don't matter
val setOpSentence = kale.SymbolDeclaration(Exp, Symbol("_ _"), ...)
val setIdentity = kale.SymbolDeclaration(Exp, Symbol(".Set"), ...)
// and so on until eventually we get to
val definition = kale.Definition(...)
// now we take the definition and try to rewrite with it
definition.step(setPatternDeclaredVeryEarly)
```

The problem is that that Kale only knows how to properly store the set once it sees the entire definition, which is after the set pattern was created. So now it would have to store the Set in a different, "default", format convert to the right implementation every time, which defeats the whole purpose of having a pattern interface. Well, once the definition is initialized, Kale could make the smarter builders available. Right, but should we expose them? Should it be in the monolithic `KaleBuilders`? In this case, `KaleBuilders` will only be able to implement Patterns for one definition and be done, because otherwise it will become ambiguous which Pattern type it constructs. And it would make the Builders a mutable object that cannot be shared between multiple threads example. And with more complex behavior. Essentially, I strongly believe we should not go on that route.

If we keep the interfaces separate, we can do the following:
```
// we could have, for convenience:
trait Builders extends OuterBuilders with PatternBuilders

// but... 
trait DefaultPatternBuilders extends PatternBuilders { ... }
class KaleBuilders extends Builders with DefaultPatternBuilders { ... }
class KaleDefinition extends outer.Definition with KalePatternBuilders { ... }

val kale: Builders  = new KaleBuilders()

// the Pattern builders are still available but they are only meant to be used while constructing the definition
val setPatternDeclaredVeryEarly = kale.Application(Symbol("_ _"), Seq(1, 2)) 
// and eventually
val definition = kale.Definition(...)

// we could still allow the following, but it would be discouraged
definition.step(setPatternDeclaredVeryEarly)

// the users would need to/should use the Definition instead
val setDefinedThroughTheDefinition = definition.Application(Symbol("_ _"), Seq(1, 2)) 
definition.step(setDefinedThroughTheDefinition)
```

Yes, `Definition` could implement `OuterBuilders` as well (or the monolithic trait), but what would that actually mean? It would not be specific to this particular `Definition` so I would argue it should not be in its interface -- the users should do `kale.` instead.

grosu commented on src/main/scala/org/kframework/kore/parser/MiniToTextToMini.scala:
```diff
@@ -0,0 +1,28 @@
+package org.kframework.kore.parser
```
We should change the names of files, too (talking about "mini").

daejunpark replied:
@lucaspena @ehildenb @msaxena2  You may want to drop this file at all. See https://github.com/kframework/kore/issues/16

grosu commented on src/main/scala/org/kframework/kore/interfaces/Interfaces.scala:
```diff
@@ -481,61 +481,251 @@ object pattern {
+    */
+  trait Outer {
+    val att: Attributes
+    val sorts: Set[p.Sort]
```
What are these `sorts`, `symbols` and `sentences` here?  The declared ones, the used ones, the error ones, etc.?  Also, they feel like utilities and not as part of the main KORE interface.  Shouldn't we keep things clean at this stage?  Maybe we should make that separation we were talking about, between a Main KORE interface and an Extended KORE interface, asap.

grosu commented on src/main/scala/org/kframework/kore/interfaces/Interfaces.scala:
```diff
@@ -481,61 +481,251 @@ object pattern {
+
+    // Derived operations
+    lazy val sorts: Set[p.Sort] = sentences.flatMap(_.sorts).toSet
+    lazy val symbols: Set[p.Symbol] = sentences.flatMap(_.symbols).toSet
```
so `symbols` collects *both* the declared and the used symbols?  For example, if one mistypes one symbol in a sentence, then that miss-typed one is also collected?  What't the use of such a method, if any?  The more I look over this code the more I feel that we should do the separation between Main and Extended KORE interfaces.  The Main one should be a mirror of the KORE language syntax that users who what to use K programmatically can use.

daejunpark commented on src/main/scala/org/kframework/kore/implementation/Kore.scala:
```diff
@@ -0,0 +1,140 @@
+
+
+object DefaultOuter {
+
```
@grosu I'm OK with this separation. It makes sense since such a separation will happens anyway (internally) in most of the backend implementations. But I also agree that this separation might be confusing to the users. So, I guess the best way to handle this is to split the interface as you suggested. The split will solve this issue as well as other issues that you mentioned below.

ehildenb replied:
@grosu So I'm not sure what you mean by eliminating the distinction between the Pattern and Outer interface. Do you mean you want to just be able to import one interface `KORE` which gets you access to all of the builders and interfaces (for `Pattern`, `Definition`, `Module`, and `Sentence`)? Or do you mean that you want to do away with the `Outer` traits altogether, in favor of implementing them in terms of ML patterns somehow?

In the first case, I think @cos mentioned a good idea where we have the separate interfaces `Pattern` and `Outer`, but have another interface `KORE` which imports both `Pattern` and `Outer`, so that we can have our cake and eat it too with the interfaces.

For the second case, if we wanted to represent the `Outer` data-structures using ML, the only encoding I can think that makes sense is using the Meta-Level. All of the `Outer` nodes in the AST would be at object level, while all the terms in rules and axioms would be at the meta-level. In that case, you can view the Scala data-structures as "one meta-level down".

Am I completely mis-understanding what you mean by "eliminate the distinction" between `Outer` and `Pattern`?

msaxena2 replied:
Another point that Lucas, Everett and I discussed in favor of keeping Pattern and Outer separate is that the AST structure with Patterns can have arbitrary depth, and thus having an AST like structure makes sense. In case of Outer, the depth will always be 3. (Node Definition,Node Module Module, Leaf Sentence).

lucaspena replied:
Regarding the first case, we just pushed up an implementation of a `KOREBuilders` interface combining `PatternBuilders` and `OuterBuilders` into one. This allows the user to import `DefaultKOREBuilders` to obtain the `DefaultPatternBuilders` and the `DefaultOuterBuilders`. If instead the user just wants `DefaultOuterBuilders`, he can just import that as well.

Currently, the way we did this has a lot of code duplication. I'm not sure if there's a simpler away using imports to avoid this, perhaps @cos can comment.

grosu replied:
@ehildenb I meant your first case.  As I said, from the perspective of one who wants to use K programmatically from Scala, the KORE API should be in a one-to-one correspondence to the KORE language.  Anything else will confuse people.  Even the very terminology, `Outer`, is confusing and adhoc.  Why `Outer`?  Radu came up with it at some moment in one of his codes, and somehow we ended up stuck with it.  I do not remember @cos' reasons for not merging the Outer and Pattern traits; can you recall them?  At this moment I think we should really separate the KORE interface into Main and Extended, as I said.  At least I can then push (hard!) for the Main interface to be faithful (identical, conceptually) to the KORE language.

grosu replied:
@lucaspena Code duplication is evil (I didn't look at your code yet)!  If you see yourself doing it it means that you are either doing the wrong thing, or doing the thing wrongly.  I think we should discuss the Main/Extended thing asap and start cleaning things up from there.  Remember that I am in charge of the KORE repository.  At this moment I do not feel good about the whole thing; that is OK, but I would like to at least feel good about the Main module.

ehildenb commented on src/main/scala/org/kframework/kore/implementation/Kore.scala:
```diff
@@ -0,0 +1,140 @@
+
+  case class Definition(modules: Seq[o.Module], att: o.Attributes) extends o.Definition {
+    override def onAttributes(f: p.Pattern => p.Pattern): Definition = Definition(modules.map(_.onAttributes(f)), att.map(f))
+  }
```
Manasvi, Lucas, and I discussed (and tried to make happen for a bit) making `Outer` implement the `AST` interface. We stored that work in another branch called `generics_ast`, where we can continue experimenting with making that work. Basically, it requires that we make `AST` (and thus `Node` and `Leaf`) parametric over its argument again. Before we had it parametric, but decided to remove that until a need arose.

This would allow us to define `onAttributes` once in the `Outer` trait, without having the code duplication in each of the implementations. Unfortunately, due to some limitations in Scala's type system regarding generics in unapply functions, it makes it harder to pattern-match directly on a `Node` and a `Leaf`. In practice, it seems that this would probably be fine.

grosu commented:
This pull request started to be outdated.  Lots of the discussions are not clearly actual anymore, such as those about the distinction between the Outer and Pattern.  
@lucaspena , can you take the lead on this and either get this ready to merge, or alternatively split it in smaller PRs?

lucaspena commented:
Looking through this pull request, it seems the brunt of it was adapted in the pull request `Main + Extended Split`. It seems the main thing originally present in this pull request that was not factored into another pull request yet was our DSL for programmatically defining KORE data structures in Scala. @ehildenb and I will work on getting this into its own pull request soon.
